Storage working group runtime changes. Iteration 2.

shamil-gadelshin opened the pull request:
## Changes
- rename "bureaucracy" to the "working group"
- rollback the 'forum-to-working group' integration
- init mint for the working group
- create runtime migrations
    - mint initialization for the storage working group
    - text constraints: `OpeningHumanReadableText`, `WorkerApplicationHumanReadableText`, `WorkerExitRationaleText`
    - remove old data: `data_directory`, `data_object_storage_registry`

shamil-gadelshin committed:
Rename "bureaucracy" module to the "working group" module

shamil-gadelshin committed:
Rollback forum changes: remove working group dependency

shamil-gadelshin committed:
Remove set storage role parameters proposal.

- remove from the codex module
- remove from the runtime proposal encode
- remove from the node chain_spec

shamil-gadelshin committed:
Remove ‘evict storage provider’ proposal.

shamil-gadelshin committed:
Remove obsolete actors module.

- remove Actors module from the runtime
- remove the actors crate

shamil-gadelshin committed:
Init storage working group mint in the storage genesis config.

shamil-gadelshin committed:
Merge branch 'nicaea' into storage_working_group_runtime_changes2

# Conflicts:
#	runtime-modules/proposals/codex/src/lib.rs
#	runtime-modules/proposals/codex/src/tests/mod.rs

shamil-gadelshin committed:
Add migration: mint for the storage working group

shamil-gadelshin committed:
Add text constraints to the storage working group

shamil-gadelshin committed:
Add text constraints runtime migration.

shamil-gadelshin committed:
Add set_mint_capacity extrinsic to the working_group module

shamil-gadelshin committed:
Add tests for the set_mint_capacity in the working group module

shamil-gadelshin committed:
Add Runtime migrations for Nicaea

mnaamani commented on node/src/chain_spec.rs:
```diff
@@ -263,10 +267,10 @@ pub fn testnet_genesis(
-        actors: Some(ActorsConfig {
-            enable_storage_role: true,
-            request_life_time: 300,
+            storage_working_group_mint_capacity: 0,
```
I don't think this is the best place for setting these configuration values.

The way it is done in substrate for the collective module which is instantiated multiple times in the runtime. You can give each instance its own configuration: 
See here https://github.com/paritytech/substrate/blob/1f536e98c509566245056369b752e9a4871de85c/bin/node/cli/src/chain_spec.rs#L284

And the collective module the logic in it's own add_extra_genesis block:
https://github.com/paritytech/substrate/blob/1f536e98c509566245056369b752e9a4871de85c/frame/collective/src/lib.rs#L146

shamil-gadelshin replied:
Actually, you're right.  I moved the code to the working group and it is much better now. Thanks for the good review.

mnaamani commented on runtime-modules/forum/src/lib.rs:
```diff
@@ -9,20 +9,19 @@
 mod mock;
 mod tests;
 
+pub use common::constraints::InputValidationLengthConstraint;
```
No need to re export, it is used in `node/src/forum_config/mod.rs` . So It would need be changed there to import from `common` instead.

shamil-gadelshin replied:
Fixed.

mnaamani commented on runtime-modules/forum/src/lib.rs:
```diff
@@ -379,14 +382,39 @@ decl_module! {
+            /*
+             * Question: when this routine is called by non sudo or with bad signature, what error is raised?
+             * Update ERror set in spec
+             */
```
```
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  DispatchError
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  20
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  0
2020-06-16 16:40:30.099 tokio-runtime-worker-6 DEBUG runtime  RequireRootOrigin
```

shamil-gadelshin replied:
I didn't change the forum code during this PR - only restored the old version. Probably that was old Bedeho's question.

mnaamani replied:
ok

mnaamani commented on runtime-modules/forum/src/lib.rs:
```diff
@@ -379,14 +382,39 @@ decl_module! {
+             * Update ERror set in spec
+             */
+
+            // Hold on to old value
```
This could be simplified a bit by `take`ing the value so you don't need to kill it later. 
```rust
// Take old value
let old_forum_sudo = <ForumSudo<T>>::take();

// Update forum sudo
if let Some(account_id) = new_forum_sudo.clone() {
    <ForumSudo<T>>::put(account_id);
}
```

shamil-gadelshin replied:
I didn't change the forum code during this PR - only restored the old version.

mnaamani commented on runtime-modules/service-discovery/src/mock.rs:
```diff
@@ -11,10 +11,10 @@ pub use sr_primitives::{
-    pub use bureaucracy::Trait;
+mod working_group_mod {
+    pub use working_group::Event;
+    pub use working_group::Instance2;
```
might make sense to alias `Instance2` to `StorageWorkingGroup` with `as` to make it easier to understand which instance we are working with? In other places as well if it makes sense, this would be important in future as we add more instances. Also why did we skip Instance1 ?

shamil-gadelshin replied:
Introduced the alias for instances. Instance1 is reserved for the Forum.

mnaamani commented on runtime-modules/storage/Cargo.toml:
```diff
@@ -73,15 +74,20 @@ default_features = false
 package = 'substrate-membership-module'
 path = '../membership'
 
+[dependencies.minting]
```
when removing the add extra genesis block out of the `data_object_storage_registry` module we can remove some unnecessary dependencies from here? Or do we still need all the dependencies of `substrate-working-group-module` as well, ie. minting, stake etc..

shamil-gadelshin replied:
Moved mint module to the dev-dependencies. All of these modules mint, stake ... etc are required for mocking.

mnaamani replied:
oh right of course :+1:

mnaamani commented on runtime-modules/storage/src/data_object_storage_registry.rs:
```diff
@@ -121,6 +125,34 @@ decl_storage! {
         /// Keeps a list of storage relationships per content id.
         pub RelationshipsByContentId get(relationships_by_content_id): map T::ContentId => Vec<T::DataObjectStorageRelationshipId>;
     }
+    add_extra_genesis {
```
This can be moved out into the working-group module itself, and the config settings can be set on each instance in the runtime configuration. see my previous comment.

shamil-gadelshin replied:
Done.

mnaamani commented on runtime-modules/storage/src/tests/data_object_storage_registry.rs:
```diff
@@ -155,3 +157,33 @@ fn test_toggle_ready() {
     });
 }
+
+#[test]
```
I think these tests will also move into the working-group module

shamil-gadelshin replied:
Agree.

mnaamani commented on runtime/src/lib.rs:
```diff
@@ -755,26 +756,11 @@ impl migration::Trait for Runtime {
-// Forum bureaucracy
-impl bureaucracy::Trait<bureaucracy::Instance1> for Runtime {
+// Storage working group
+impl working_group::Trait<working_group::Instance2> for Runtime {
```
Why the change to Instance2 ? Are you reserving Instance1 for the content working group? Or was instance1 for the forum.. I kind of got confused here. Maybe we can import working_group::Instance2 and working_group::Instance1 with an alias at the top to avoid this confusion and possibly making a mistake references the wrong instance.

shamil-gadelshin replied:
Introduced the alias for instances. Instance1 is reserved for the Forum.

mnaamani commented on runtime/src/lib.rs:
```diff
@@ -867,7 +853,6 @@ construct_runtime!(
         Memo: memo::{Module, Call, Storage, Event<T>},
         Members: members::{Module, Call, Storage, Event<T>, Config<T>},
         Forum: forum::{Module, Call, Storage, Event<T>, Config<T>},
-        Actors: actors::{Module, Call, Storage, Event<T>, Config},
```
Good bye 👍

mnaamani commented on runtime/src/lib.rs:
```diff
@@ -884,9 +869,9 @@ construct_runtime!(
-        ForumBureaucracy: bureaucracy::<Instance1>::{Module, Call, Storage, Event<T>},
-        StorageBureaucracy: bureaucracy::<Instance2>::{Module, Call, Storage, Event<T>},
+        // --- Working groups
+        // reserved for the future use: ForumWorkingGroup: working_group::<Instance1>::{Module, Call, Storage, Event<T>},
```
oh hello :) that explains things.

mnaamani commented on runtime-modules/storage/src/data_object_storage_registry.rs:
```diff
@@ -121,6 +125,34 @@ decl_storage! {
+            if let Ok(mint_id) = mint_id_result {
+                <working_group::Mint::<T, working_group::Instance2>>::put(mint_id);
+            } else {
+                print("Failed to create a mint for the storage working group");
```
I would add a panic! here there is no point initializing a chain spec with bad state. And this isn't runtime code so its okay to fail loudly.

shamil-gadelshin replied:
Ok.

mnaamani commented on runtime/src/migration.rs:
```diff
@@ -18,10 +21,24 @@ impl<T: Trait> Module<T> {
+
+        // ** Order is important!
+        Self::clear_storage_data_object_registry_data();
+        Self::clear_storage_data_directory_data();
```
maybe both steps in a single `fn clear_storage_data()` and avoid need for the "order is important" comment ?

How about also resetting the `next_*_id` state variables to their initial values as well ?

shamil-gadelshin replied:
Both fixed.

mnaamani commented on runtime-modules/storage/src/data_object_storage_registry.rs:
```diff
@@ -25,10 +25,14 @@
 use sr_primitives::traits::{MaybeSerialize, Member, SimpleArithmetic};
-use srml_support::{decl_error, decl_event, decl_module, decl_storage, ensure, Parameter};
+use srml_support::{
+    decl_error, decl_event, decl_module, decl_storage, ensure, print, Parameter, StorageValue,
```
adding `StorageValue` seems un-necessary. Did clippy complain?

shamil-gadelshin replied:
Rust thing: a trait declaration should be in scope.

mnaamani replied:
correct, but if I'm not mistaken the `decl_storage!` macro now brings the necessary ones into scope.. like StorageMap, StorageValue etc.. anyway its not a big deal if it stays.

shamil-gadelshin replied:
Hmm. You are right. It compiles without it. There were errors previously.

mnaamani reviewed (changes requested):
Fantastic work.
left a few suggestions.

shamil-gadelshin committed:
Fix review comments

mnaamani commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1247,6 +1261,30 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
+        opening_human_readable_text_constraint: InputValidationLengthConstraint,
+        worker_application_human_readable_text_constraint: InputValidationLengthConstraint,
+        worker_exit_rationale_text_constraint: InputValidationLengthConstraint,
+        storage_working_group_mint_capacity: minting::BalanceOf<T>,
```
```suggestion
        working_group_mint_capacity: minting::BalanceOf<T>,
```

shamil-gadelshin replied:
Fixed.

mnaamani commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1247,6 +1261,30 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
+    ) {
+        // Create a mint.
+        let mint_id_result =
+            <minting::Module<T>>::add_mint(storage_working_group_mint_capacity, None);
```
```suggestion
            <minting::Module<T>>::add_mint(working_group_mint_capacity, None);
```

mnaamani commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1247,6 +1261,30 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
+        if let Ok(mint_id) = mint_id_result {
+            <Mint<T, I>>::put(mint_id);
+        } else {
+            panic!("Failed to create a mint for the storage working group");
```
```suggestion
            panic!("Failed to create a mint for the working group");
```

mnaamani replied:
I'm not sure if we can identify which instance also in the error message?

mnaamani commented on runtime-modules/working-group/src/tests/mock.rs:
```diff
@@ -124,17 +124,42 @@ impl recurringrewards::Trait for Test {
+pub type TestWorkingGroupInstance = crate::Instance1;
+pub type TestWorkingGroup = Module<Test, TestWorkingGroupInstance>;
+
+pub(crate) const STORAGE_WORKING_GROUP_MINT_CAPACITY: u64 = 40000;
```
`STORAGE_` prefix for these constants here could be removed.. I'm just being picky :)

shamil-gadelshin replied:
Fixed.

mnaamani commented on runtime-modules/working-group/src/tests/mod.rs:
```diff
@@ -1592,8 +1617,37 @@ fn set_working_group_mint_capacity_fails_with_mint_not_found() {
 }
+
+#[test]
+fn ensure_setting_genesis_storage_working_group_mint_succeeds() {
```
`_storage` in the names of the test for the working group can be dropped.

shamil-gadelshin replied:
Fixed.

mnaamani commented on runtime/src/migration.rs:
```diff
@@ -117,9 +115,10 @@ impl<T: Trait> Module<T> {
-    }
 
-    fn clear_storage_data_directory_data() {
+        storage::data_object_storage_registry::NextRelationshipId::<T>::put(<T as storage::data_object_storage_registry::Trait>::DataObjectStorageRelationshipId::one());
```
`one()` works because the constant `DEFAULT_FIRST_RELATIONSHIP_ID` defined in the module is equal to 1.
But maybe better to just`kill()` the value, next time it is read it will use the default storage value in this case configured as the constant:

```rust
pub FirstRelationshipId get(first_relationship_id) config(first_relationship_id):
            T::DataObjectStorageRelationshipId = T::DataObjectStorageRelationshipId::from(DEFAULT_FIRST_RELATIONSHIP_ID);
```

shamil-gadelshin replied:
Fixed.

mnaamani committed:
proper use of mint in working groups

- All working groups: Ensure mint exists before filling opening if a reward policy is used
  because mutation of filled positions can happen followed by a failed check for the mint.
- working-group module: check mint value exists before reading from storage.

mnaamani committed:
content-working-group: check mint value is set before using it

shamil-gadelshin committed:
Fix review comments

shamil-gadelshin committed:
Disable storage data migration

mnaamani committed:
mint: do not do a bug-check of storage value being set

mnaamani commented on runtime-modules/proposals/codex/src/proposal_types/mod.rs:
```diff
@@ -47,14 +46,8 @@ pub enum ProposalDetails<MintedBalance, CurrencyBalance, BlockNumber, AccountId,
     SetContentWorkingGroupMintCapacity(MintedBalance),
 
-    /// AccountId for the `evict storage provider` proposal
-    EvictStorageProvider(AccountId),
```
I just realized we actually need to keep a "dummy" value here to preserve the enumeration index correct when reading state of proposals of this type. 

can we change it to:
`_EvicStorageProvider(AccountId)`

shamil-gadelshin replied:
Fixed.

mnaamani commented on runtime-modules/proposals/codex/src/proposal_types/mod.rs:
```diff
@@ -47,14 +46,8 @@ pub enum ProposalDetails<MintedBalance, CurrencyBalance, BlockNumber, AccountId,
     SetValidatorCount(u32),
-
-    /// Role parameters for the `set storage role parameters` proposal
-    SetStorageRoleParameters(RoleParameters<CurrencyBalance, BlockNumber>),
```
`_SetStorageRoleParameters(RoleParameters<CurrencyBalance, BlockNumber>)`
This means we also need to preserve the `struct RoleParameters` type.

I think we can remove the encapsulated value if we can ensure we don't try to read in the runtime any proposal detail of this variant.

shamil-gadelshin replied:
Fixed.

shamil-gadelshin committed:
Merge pull request #4 from mnaamani/fill-opening-bug-fix

Fix use of mint in working groups

shamil-gadelshin committed:
Comment unused code

shamil-gadelshin committed:
Restore the deleted proposal details to preserve pioneer compatibility

shamil-gadelshin committed:
Update rust version and fix clippy issues
